File: frontend/src/services/categoryService.ts
import { apiClient } from './apiClient';
import { Category } from '../types';

export interface CategoryRequest {
  name: string;
  description?: string;
}

export const categoryService = {
  async getAll(): Promise<Category[]> {
    const res = await apiClient.get<Category[]>('/categories');
    return res.data;
  },

  async create(data: CategoryRequest): Promise<Category> {
    const res = await apiClient.post<Category>('/categories', data);
    return res.data;
  },

  async update(id: number, data: CategoryRequest): Promise<Category> {
    const res = await apiClient.put<Category>(`/categories/${id}`, data);
    return res.data;
  },

  async delete(id: number): Promise<void> {
    await apiClient.delete(`/categories/${id}`);
  },
};

File: backend/src/main/java/com/bookstore/category/service/CategoryService.java
package com.bookstore.category.service;

import com.bookstore.book.repository.BookRepository;
import com.bookstore.category.dto.CategoryRequest;
import com.bookstore.category.entity.Category;
import com.bookstore.category.repository.CategoryRepository;
import java.util.List;
import org.springframework.stereotype.Service;

@Service
public class CategoryService {
  private final CategoryRepository categories;
  private final BookRepository books;

  public CategoryService(CategoryRepository categories, BookRepository books) {
    this.categories = categories;
    this.books = books;
  }

  public List<Category> all() {
    return categories.findAll();
  }

  public Category create(CategoryRequest request) {
    return save(new Category(), request);
  }

  public Category update(Long id, CategoryRequest request) {
    return save(categories.findById(id).orElseThrow(), request);
  }

  public void delete(Long id) {
    if (books.existsByCategoryId(id)) {
      throw new IllegalStateException("Cannot delete a category linked to books");
    }
    categories.deleteById(id);
  }

  private Category save(Category category, CategoryRequest request) {
    category.setName(request.name().trim());
    category.setDescription(request.description());
    return categories.save(category);
  }
}